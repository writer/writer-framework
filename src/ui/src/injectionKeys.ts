import { ComputedRef, InjectionKey, Ref, VNode } from "vue";
import {
	BuilderManager,
	CollaborationManager,
	Component,
	Core,
	InstancePath,
	InstancePathItem,
	NotesManager,
	SecretsManager,
} from "./writerTypes";

export default {
	core: Symbol() as InjectionKey<Core>,
	builderManager: Symbol() as InjectionKey<BuilderManager>,
	notesManager: Symbol() as InjectionKey<NotesManager>,
	collaborationManager: Symbol() as InjectionKey<CollaborationManager>,
	secretsManager: Symbol() as InjectionKey<SecretsManager>,
	evaluatedFields: Symbol() as InjectionKey<Record<string, ComputedRef<any>>>,
	componentId: Symbol() as InjectionKey<Component["id"]>,
	isBeingEdited: Symbol() as InjectionKey<Ref<boolean>>,
	isDisabled: Symbol() as InjectionKey<Ref<boolean>>,
	getChildrenVNodes: Symbol() as InjectionKey<
		(instanceNumber?: InstancePathItem["instanceNumber"]) => VNode[]
	>,
	renderProxiedComponent: Symbol() as InjectionKey<
		(
			componentId: Component["id"],
			instanceNumber: InstancePathItem["instanceNumber"],
			ext?: { class?: string[]; contextSlot?: string },
		) => VNode
	>,
	instancePath: Symbol() as InjectionKey<InstancePath>,
	flattenedInstancePath: Symbol() as InjectionKey<string>,
	instanceData: Symbol() as InjectionKey<Ref[]>,
};
