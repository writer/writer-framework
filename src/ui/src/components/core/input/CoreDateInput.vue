<template>
	<BaseInputWrapper
		ref="rootInstance"
		:label="fields.label.value"
		class="CoreDateInput"
	>
		<input
			type="date"
			:value="formValue"
			@change="
				($event) =>
					handleInput(
						($event.target as HTMLInputElement).value,
						'wf-date-change',
					)
			"
		/>
	</BaseInputWrapper>
</template>

<script lang="ts">
import { FieldType } from "@/writerTypes";
import { cssClasses } from "@/renderer/sharedStyleFields";
import BaseInputWrapper from "../base/BaseInputWrapper.vue";
import { ComponentPublicInstance } from "vue";

const description =
	"A user input component that allows users to select a date using a date picker interface.";

const onChangeHandlerStub = `
def onchange_handler(state, payload):

	# Set the state variable "new_date" to the new value, provided as a YYYY-MM-DD string.

	state["new_date"] = payload`;

export default {
	writer: {
		name: "Date Input",
		description,
		category: "Input",
		fields: {
			label: {
				name: "Label",
				init: "Input Label",
				type: FieldType.Text,
			},
			cssClasses,
		},
		events: {
			"wf-date-change": {
				desc: "Capture changes to this control.",
				stub: onChangeHandlerStub,
				bindable: true,
				eventPayloadExample: new Date().toISOString().split("T")[0],
			},
		},
	},
};
</script>
<script setup lang="ts">
import { inject, ref } from "vue";
import injectionKeys from "@/injectionKeys";
import { useFormValueBroker } from "@/renderer/useFormValueBroker";

const fields = inject(injectionKeys.evaluatedFields);
const rootInstance = ref<ComponentPublicInstance | null>(null);
const wf = inject(injectionKeys.core);
const instancePath = inject(injectionKeys.instancePath);

const { formValue, handleInput } = useFormValueBroker(
	wf,
	instancePath,
	rootInstance,
);
</script>

<style scoped>
@import "@/renderer/sharedStyles.css";
@import "@/renderer/colorTransformations.css";

.CoreDateInput {
	width: fit-content;
}

input {
	width: 100%;
	max-width: 20ch;
	margin: 0;
	border: 1px solid var(--separatorColor);
	border-radius: 8px;
	padding: 8.5px 12px 8.5px 12px;
	font-size: 0.875rem;
	outline: none;
	cursor: text;
}

input:focus {
	border: 1px solid var(--softenedAccentColor);
	box-shadow: 0px 0px 0px 3px rgba(81, 31, 255, 0.05);
}
</style>
