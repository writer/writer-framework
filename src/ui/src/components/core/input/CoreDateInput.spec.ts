import { describe, expect, it, vi } from "vitest";

import CoreDateInput from "./CoreDateInput.vue";
import { flushPromises, mount } from "@vue/test-utils";
import injectionKeys from "@/injectionKeys";
import { ref } from "vue";
import { buildMockComponent, buildMockCore, mockProvides } from "@/tests/mocks";

describe("CoreDateInput", () => {
	it("should render value from the state and forward emit", async () => {
		const { core, userState } = buildMockCore();
		userState.value = { key: "2024-12-14" };
		core.addComponent(
			buildMockComponent({
				handlers: { "wf-date-change": "python_handler" },
				binding: {
					eventType: "",
					stateRef: "key",
				},
			}),
		);

		const wrapper = mount(CoreDateInput, {
			global: {
				provide: {
					...mockProvides,
					[injectionKeys.core as symbol]: core,
					[injectionKeys.evaluatedFields as symbol]: {
						label: ref("This is the label"),
					},
				},
			},
		});

		await flushPromises();

		const input = wrapper.get("input");
		expect(input.element.value).toStrictEqual("2024-12-14");

		const dispatchEvent = vi.spyOn(wrapper.vm.$el, "dispatchEvent");

		await input.trigger("change");

		await flushPromises();

		expect(dispatchEvent).toHaveBeenCalledOnce();

		expect(wrapper.element).toMatchSnapshot();
	});
});
