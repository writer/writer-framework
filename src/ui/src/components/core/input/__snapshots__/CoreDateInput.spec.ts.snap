// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CoreDateInput > should render value from the state and forward emit 1`] = `
<div
  class="BaseInputWrapper CoreDateInput"
  data-v-7e12caaf=""
  data-v-acf95c76=""
>
  <label
    data-v-7e12caaf=""
  >
    This is the label
  </label>
  
  <input
    data-v-acf95c76=""
    type="date"
  />
  
</div>
`;
